src/price_data.py

```diff
@@ -267,33 +268,49 @@ def _get_price_bitpanda_pro(
         # if there were no trades in the requested time frame, the
         # returned data will be empty
         for t in timeframes:
-            end = utc_time
-            begin = utc_time - datetime.timedelta(minutes=t)
+            num_offsets = 0
+            for num_offsets in range(6):
+                # if no trades can be found, move 30 min window to the past
+                window_offset = num_offsets * 30
+                end = utc_time.astimezone(datetime.timezone.utc) \
+                    - datetime.timedelta(minutes=window_offset)
+                begin = end - datetime.timedelta(minutes=t)
+                if num_offsets:
+                    log.warning(
+                        f"No price data found for {base_asset} / {quote_asset} "
+                        f"at {end}, moving {timeframes[-1]} minute window to the past."
+                    )
+
+                # https://github.com/python/mypy/issues/3176
+                params: dict[str, Union[int, str]] = {
+                    "unit": "MINUTES",
+                    "period": t,
+                    # convert ISO 8601 format to RFC3339 timestamp
+                    "from": begin.isoformat().replace('+00:00', 'Z'),
+                    "to": end.isoformat().replace('+00:00', 'Z'),
+                }
+                log.debug(
+                    f"Calling Bitpanda API for {base_asset} / {quote_asset} "
+                    f"price for {t} minute timeframe ending at {end}"
+                )
+                r = requests.get(baseurl, params=params)
 
-            # https://github.com/python/mypy/issues/3176
-            params: dict[str, Union[int, str]] = {
-                "unit": "MINUTES",
-                "period": t,
-                "from": begin.isoformat(),
-                "to": end.isoformat(),
-            }
-            r = requests.get(baseurl, params=params)
+                assert r.status_code == 200, "No valid response from Bitpanda API"
+                data = r.json()
 
-            assert r.status_code == 200
+                # exit loop if data is valid or maximum timeframe is not reached yet
+                if data or t != timeframes[-1]:
```

Comment:
Reading this from top to bottom, I thought at first, that we offset the window for each timeframe and got confused.

What do you think of this approach? It might be easier to understand when reading the code. We can update the comment above the for-loop to match our code.

```
# Try to find the price in the most detailed timeframe, to get the best matching price for the transaction.
# If we can not find the price in a timeframe, use the next bigger frame and try again.
# If we reached the highest timeframe, move the fetched time window into the past, to get the latest transaction from the API.
for t in timeframes:
  # Maximum number of allowed offsets into the past to find a valid price before we throw an error.
  # We do not offset the time window as long as we can choose a bigger timeframe instead.
  num_offsets = 6 if t == timeframes[-1] else 1
  for num_offset in range(num_offsets):
    ## attention: i changed the for-loop-variable name to singular
    ...
```

Reply:
yes, that's easier to read, I made the changes